api/configuration/v1alpha1/kongkey_types.go

```diff
@@ -0,0 +1,182 @@
+func (t *KongKey) GetKonnectID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ID
+}
+
+// SetKonnectID sets the Konnect ID in the KongKey status.
+func (t *KongKey) SetKonnectID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ID = id
+}
+
+// GetControlPlaneID returns the ControlPlane ID in the KongKey status.
+func (t *KongKey) GetControlPlaneID() string {
+	if t.Status.Konnect == nil {
+		return ""
+	}
+	return t.Status.Konnect.ControlPlaneID
+}
+
+// SetControlPlaneID sets the ControlPlane ID in the KongKey status.
+func (t *KongKey) SetControlPlaneID(id string) {
+	if t.Status.Konnect == nil {
+		t.initKonnectStatus()
+	}
+	t.Status.Konnect.ControlPlaneID = id
+}
+
+// GetTypeName returns the KongKey Kind name.
+func (t KongKey) GetTypeName() string {
+	return "KongKey"
+}
+
+// GetConditions returns the Status Conditions.
+func (t *KongKey) GetConditions() []metav1.Condition {
+	return t.Status.Conditions
+}
+
+// SetConditions sets the Status Conditions.
+func (t *KongKey) SetConditions(conditions []metav1.Condition) {
+	t.Status.Conditions = conditions
+}
+
+func (t *KongKey) initKonnectStatus() {
+	t.Status.Konnect = &konnectv1alpha1.KonnectEntityStatusWithControlPlaneAndKeySetRef{}
+}
+
+// KongKeySpec defines the spec for a KongKey.
+type KongKeySpec struct {
+	// ControlPlaneRef is a reference to a Konnect ControlPlane this KongKey is associated with.
+	// +optional
+	ControlPlaneRef *ControlPlaneRef `json:"controlPlaneRef,omitempty"`
+
+	// KeySetRef is a reference to a KeySet this KongKey is attached to.
+	// +optional
+	KeySetRef *KeySetRef `json:"keySetRef,omitempty"`
+
+	// KongKeyAPISpec are the attributes of the KongKey itself.
+	KongKeyAPISpec `json:",inline"`
+}
+
+// KongKeyAPISpec defines the attributes of a Kong Key.
+// +kubebuilder:validation:XValidation:rule="has(self.jwk) || has(self.pem)", message="Either 'jwk' or 'pem' must be set"
+type KongKeyAPISpec struct {
+	// KID is a unique identifier for a key.
+	// When JWK is provided, KID has to match the KID in the JWK.
+	// +kubebuilder:validation:MinLength=1
+	KID string `json:"kid"`
+
+	// Name is an optional name to associate with the given key.
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// JWK is a JSON Web Key represented as a string.
+	// The JWK must contain a KID field that matches the KID in the KongKey.
+	// Either JWK or PEM must be set.
+	// +optional
+	JWK *string `json:"jwk,omitempty"`
+
+	// PEM is a keypair in PEM format.
+	// Either JWK or PEM must be set.
+	// +optional
+	PEM *PEMKeyPair `json:"pem,omitempty"`
+
+	// Tags is an optional set of strings associated with the Key for grouping and filtering.
+	// +optional
+	Tags []string `json:"tags,omitempty"`
+}
+
+// PEMKeyPair defines a keypair in PEM format.
+type PEMKeyPair struct {
+	// The private key in PEM format.
+	// +kubebuilder:validation:MinLength=1
+	PrivateKey string `json:"private_key"`
```

Comment:
Is there any requirement to redact the `PrivateKey`?

Reply:
It's one of the many fields that fall into the sensitive data bucket. I've just opened a discussion in the project's Slack channel so we can make a decision regarding them.

Reply:
This is the same as for other sensitive objects.

Given that we have separate types, users can enforce RBAC based on those.

Reply:
Alright, let's proceed with that, at least for now. 👍